tokio/src/util/slab.rs

```diff
@@ -0,0 +1,748 @@
+            let mut slots = match page.slots.try_lock() {
+                Ok(slots) => slots,
+                // If the lock cannot be acquired due to being held by another
+                // thread, don't try to compact the page.
+                _ => continue,
+            };
+
+            if slots.used > 0 || slots.slots.capacity() == 0 {
+                // The page is in use or it has not yet been allocated. Either
+                // way, there is no more work to do.
+                continue;
+            }
+
+            // Remove the slots vector from the page. This is done so that the
+            // freeing process is done outside of the lock's critical section.
+            let vec = mem::replace(&mut slots.slots, vec![]);
+            slots.head = 0;
+
+            // Drop the vec outside of the lock
+            drop(slots);
+
+            // Clear cache
+            self.cached[idx].slots = ptr::null();
+            self.cached[idx].init = 0;
+
+            drop(vec);
+        }
+    }
+}
+
+impl<T> fmt::Debug for Slab<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("Slab").finish()
+    }
+}
+
+impl<T: Entry> Allocator<T> {
+    /// Allocate a new entry and return a handle to the entry.
+    ///
+    /// Scans pages from smallest to biggest, stopping when a slot is found.
+    /// Pages are allocated if necessary.
+    ///
+    /// Returns `None` if the slab is full.
+    pub(crate) fn allocate(&self) -> Option<(Address, Ref<T>)> {
+        // Find the first available slot.
+        for page in &self.pages[..] {
+            if let Some((addr, val)) = Page::allocate(page) {
+                return Some((addr, val));
+            }
+        }
+
+        None
+    }
+}
+
+impl<T> fmt::Debug for Allocator<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("slab::Allocator").finish()
+    }
+}
+
+impl<T> ops::Deref for Ref<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        // Safety: `&mut` is never handed out to the underlying value. The page
+        // is not freed until all `Ref` values are dropped.
+        unsafe { &(*self.value).value }
+    }
+}
+
+impl<T> Drop for Ref<T> {
+    fn drop(&mut self) {
+        // Safety: `&mut` is never handed out to the underlying value. The page
+        // is not freed until all `Ref` values are dropped.
+        let _ = unsafe { (*self.value).release() };
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for Ref<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (**self).fmt(fmt)
+    }
+}
+
+impl<T: Entry> Page<T> {
+    // Allocates an object, returns the ref and address.
+    //
+    // `self: &Arc<Page<T>>` is avoided here as this would not work with the
+    // loom `Arc`.
+    fn allocate(me: &Arc<Page<T>>) -> Option<(Address, Ref<T>)> {
+        // Before acquiring the lock, use the `used` hint.
+        if me.used.load(Acquire) == me.len {
```

Comment:
As I understand orderings, there's no reason to use anything but `Relaxed` unless you need to synchronize _other_ data using the atomic (e.g. as a lock).

Reply:
You are correct.

Reply:
This isn't quite right; in order to establish a happens-after relationship between a concurrent load and modification, it is necessary to use an `Acquire`/`Release` pair. See https://doc.rust-lang.org/nomicon/atomics.html#acquire-release

It just so happens that on x86_64 CPUs, all memory accesses are essentially _already_ `AcqRel`, so memory orderings really only effect compiler reordering. However, on other architectures such as ARM, the hardware memory ordering is weaker --- so it's possible to write code that behaves correctly on x86 but is invalid under the memory model, and thus will not behave correctly on weakly-ordered machines.

Reply:
I switched.